mon-entreprise/source/components/utils/useSearchParamsSimulationSharing.ts

```diff
@@ -0,0 +1,175 @@
+// XXX Filter-out search params starting with something non-alphabetic & change
+// _ to targetUnit. To prevent collision with other services, which will have to
+// be modified to start with _ for example.
+import { useEffect, useState } from 'react'
+import { useSelector, useDispatch } from 'react-redux'
+import { SimulationConfig, Situation } from 'Reducers/rootReducer'
+import { useSearchParams } from 'Components/utils/useSearchParams'
+import { useEngine } from 'Components/utils/EngineContext'
+import {
+	configSelector,
+	situationSelector,
+	targetUnitSelector,
+} from 'Selectors/simulationSelectors'
+import { ParsedRules } from 'publicodes'
+import { DottedName } from 'modele-social'
+import { updateSituation, updateUnit, setActiveTarget } from 'Actions/actions'
+
+type Objectifs = (string | { objectifs: string[] })[]
+
+const objectifsOfConfig = (config: Partial<SimulationConfig>) =>
+	(config.objectifs as Objectifs).flatMap((objectifOrSection) => {
+		if (typeof objectifOrSection === 'string') {
+			return [objectifOrSection]
+		}
+		return objectifOrSection.objectifs
+	})
+
+export default function useSearchParamsSimulationSharing() {
+	const [urlSituationIsExtracted, setUrlSituationIsExtracted] = useState(false)
+	const [searchParams, setSearchParams] = useSearchParams()
+	const config = useSelector(configSelector)
+	const situation = useSelector(situationSelector)
+	const targetUnit = useSelector(targetUnitSelector)
+	const engine = useEngine()
+	const dispatch = useDispatch()
+
+	useEffect(() => {
+		const hasConfig = Object.keys(config).length > 0
+		if (!hasConfig) return
+
+		const objectifs = objectifsOfConfig(config)
+		// On load:
+		if (!urlSituationIsExtracted) {
+			const {
+				situation: newSituation,
+				targetUnit: newTargetUnit,
+			} = getSituationFromSearchParams(searchParams, engine.getParsedRules())
+			if (newTargetUnit) dispatch(updateUnit(newTargetUnit))
+			Object.entries(newSituation).forEach(([dottedName, value]) => {
+				dispatch(updateSituation(dottedName as DottedName, value))
+			})
+			const newActiveTarget = Object.keys(newSituation).filter((dottedName) =>
+				objectifs.includes(dottedName)
+			)[0]
+			if (newActiveTarget) {
+				dispatch(setActiveTarget(newActiveTarget as DottedName))
+			}
+			setUrlSituationIsExtracted(true)
+
+			setSearchParams('')
+			return
+		}
+	}, [
+		dispatch,
+		config,
+		engine,
+		searchParams,
+		setSearchParams,
+		situation,
+		targetUnit,
+		urlSituationIsExtracted,
+	])
+
+	return () =>
+		getSearchParamsFromSituation(
+			situation,
+			targetUnit || (config['unité par défaut'] as string),
+			engine.getParsedRules()
+		)
+}
+
+// Simple (de)serialization based on value type
+export const serialize = (
+	value: Record<string, unknown> | string | number
+): string => {
+	switch (typeof value) {
+		case 'object':
+			return `${JSON.stringify(value)}`
+		case 'string':
+			return value
+		case 'number':
+			return `_${value}`
```

Comment:
Dans 80% des cas, c'est des boolean ou des nombre qui sont retournée. Je pense qu'il faudrait être un peu plus subtil pour la sérialisation / deserialisation : c'est quelque chose qu'on ne pourra pas changer par la suite...

Je pense que le mieux c'est de se baser sur l'évaluation de la situation dans le contexte courant, plutôt que de l'objet sauvegardé.

Pour être plus clair : lorsqu'on clique sur "SMIC" la valeur brut dans la situation vaut 'contrat salarié . SMIC' et non '1539€'.

Du coup pour le partage de la situation je serai d'avis d'évaluer cette valeur dans le contexte et de retourner la valeur évaluée. 

On aurait : 
1. `engine.evaluate(value)` qui retourne un EvaluatedNode
2. Serialisation de la valeur en utilisant le même code que : https://github.com/betagouv/mon-entreprise/blob/master/publicodes/source/components/rule/RuleSource.tsx#L110 

A ta dispo pour en discuter

Reply:
> Pour être plus clair : lorsqu'on clique sur "SMIC" la valeur brut dans la situation vaut 'contrat salarié . SMIC' et non '1539€'.

Pour être franc ce n'est pas plus clair =) Je ne vois pas de quoi on parle: les valeurs de la situation ne sont-elle pas des EvaluatedNode? Dans ce cas ce sont des quoi?  
C'est justement ce que je me suis "juré" de ne pas regarder dans cette PR, car il s'agit de commencer à regarder le moteur plus en détail.

> Serialisation de la valeur en utilisant le même code que : https://github.com/betagouv/mon-entreprise/blob/master/publicodes/source/components/rule/RuleSource.tsx#L110

Comme indiqué à ce fichier par un commentaire et comme indiqué par moi dans un autre commentaire qque part, c'est le rôle de l'Engine de gérer cela.  
Ceci dit cela ne semble pas non plus trop compliqué à ré-intégrer dans l'engine. Il faudrait que je prenne le temps de comprendre ce que ça veut dire.

> 'est quelque chose qu'on ne pourra pas changer par la suite

A mon avis c'est point crucial de ton raisonnement. Est-ce si vrai que cela? Si on ne prévoit pas d'écrire ces liens en dur quelque part, ça devrait aller. Par ex on peut ajouter une clé de version dans les paramètres, et si jamais la version est inférieure à la version courante, on s'excluse platement à l'utilisateur (voire même on garde 2 versions du code et on décommissionne plus tard, why not).

Reply:
> Pour être franc ce n'est pas plus clair =) Je ne vois pas de quoi on parle: les valeurs de la situation ne sont-elle pas des EvaluatedNode? Dans ce cas ce sont des quoi?

Ce sont des ASTNode tout court, parsés mais non évalués. Ils sont évalués lors de l'évaluation, ce qui permet d'avoir des valeurs dépendante d'autres valeurs dans la situation (comme le smic qui est actualisé en fonction du nombre d'heures travaillées au temps partiel).

> Comme indiqué à ce fichier par un commentaire et comme indiqué par moi dans un autre commentaire qque part, c'est le rôle de l'Engine de gérer cela.

Oui, tout à fait d'où réutiliser la logique et la déplacer au bon endroit (l'engine).

> Par ex on peut ajouter une clé de version dans les paramètres, et si jamais la version est inférieure à la version courante, on s'excluse platement à l'utilisateur

C'est un peu dommage d'avoir ça comme horizon pour les utilisateurs qui partagent. Et ça pourra être compliqué d'ajouter une couche pour la compatibilité plus tard. Après, c'est mon avis, je pense que c'est bien moins couteux de modifier maintenant que plus tard. Je peux me pencher dessus si tu veux.